### ISSUE TITLE
BUG: env.toctree_includes different between parallel and serial builds

### ISSUE THREAD

**Issue by larsoner** [sphinx-doc/sphinx#12409]
### Describe the bug

In SciPy we had a bug where using pydata-sphinx-theme in serial builds was fine, but in parallel builds a `jinja` `context` did some incorrect HTML replacement. The short version is that the PST function looks for the ancestors of a given page using [`_get_toctree_ancestors(env.toctree_includes, pagename)`](https://github.com/sphinx-doc/sphinx/blob/f0dc8239cd291176f2421d08f2b3c096cb5785c1/sphinx/environment/adapters/toctree.py#L478) and decides based on the ancestors what HTML to show.

In the correct-behavior case, **which always happens in serial mode and sometimes happens in parallel mode**, for the [`reference/cluster.hierarchy` page](https://github.com/scipy/scipy/blob/main/doc/source/reference/cluster.hierarchy.rst?plain=1) we get what you'd expect:
```
ancestors=['reference/cluster.hierarchy', 'reference/cluster', 'reference/index']
```
In the incorrect case, which happens *non-deterministically* in parallel builds only, we get:
```
ancestors=['reference/cluster.hierarchy', 'reference/cluster', 'reference/index', 'dev/contributor/contributor_toc', 'dev/index']
```
You can see the addition of `'dev/contributor/contributor_toc'` -- this isn't entirely unreasonable because `reference/index` *is* [added to a hidden toctree of that page](https://github.com/scipy/scipy/blob/9d543aec2fefdf14daaae459ca00e3d03af61ac5/doc/source/dev/contributor/contributor_toc.rst?plain=1#L97). But it does deviate from the serial behavior to have those in there.

I wonder if it's perhaps due to when the forked reading subprocesses return, and which `env` gets merged into which (and what takes precedence). There is also some `setdefault` code that could be involved. I'm not sure 🤷 

### How to Reproduce

After following SciPy's dev docs:
```
python dev.py doc -j6
```
or if you don't want it to build but have SciPy and pydata-sphinx-theme installed, clone SciPy `main` (before https://github.com/scipy/scipy/pull/20897 gets merged, like on https://github.com/scipy/scipy/commit/9d543aec2fefdf14daaae459ca00e3d03af61ac5) and then enter the `doc` dir and do:
```
python -msphinx -WT --keep-going -b html -d build/doctrees -E source build/html -j6
```
it will complain about bad refs and stuff but then if you inspect the output you might see the incorrect behavior (left) or the correct behavior (right) 🎲 :

| Incorrect | Correct |
| -- | -- |
| <img alt="Screenshot 2024-06-04 at 12 31 30 PM" src="https://github.com/sphinx-doc/sphinx/assets/2365790/900802fb-9964-46ae-a383-f29d7239e992"> | <img alt="Screenshot 2024-06-04 at 12 16 30 PM" src="https://github.com/sphinx-doc/sphinx/assets/2365790/27e93178-3a1b-47a9-9d87-b49001e92d3b"> |

### Environment Information

```text
Tried on latest `master` and Sphinx 5.3. Replicated on macOS but originally seen on Linux in CIs. Tried with Python 3.11 and 3.12, etc.
```

### Sphinx extensions

```python
I commented out a few of them to speed up the build but ultimately used:

    'sphinx.ext.autodoc',
    'sphinx.ext.autosummary',
    'sphinx.ext.coverage',
    'sphinx.ext.mathjax',
    'sphinx.ext.intersphinx',
    'numpydoc',
    'sphinx_design',
    'doi_role',
    'matplotlib.sphinxext.plot_directive',
```

**Comment by picnixz** [comment]
> Tried on latest `master` and Sphinx 5.3. 

So the bug has been since 5.3? I'm really surprised that no one noticed... Just wondering but.. what happens if you try using Python 3.9 and Python 3.10? If the same happens, I think we might have an issue with the merge environment process. By the way, does it also happen using another theme?

I'd like a MWE (with at least 10 documents because otherwise you cannot really trigger the parallel build) which does not use any custom extensions so that I can know which are the side-effects.

**Comment by jayaddison** [comment]
Adding a note to mention that I'm currently investigating a very similar behaviour -- perhaps a duplicate -- in #6714.
